scalding-core/src/main/scala/com/twitter/scalding/typed/TypedPipe.scala

```diff
@@ -149,29 +150,34 @@ object TypedPipe extends Serializable {
           case ComputedValue(pipe) => CrossPipe(left, pipe)
         }
    }
-   final case class DebugPipe[T](pipe: TypedPipe[T]) extends TypedPipe[T]
+   final case class DebugPipe[T](input: TypedPipe[T]) extends TypedPipe[T]
    final case class FilterKeys[K, V](input: TypedPipe[(K, V)], fn: K => Boolean) extends TypedPipe[(K, V)]
    final case class Filter[T](input: TypedPipe[T], fn: T => Boolean) extends TypedPipe[T]
-   final case class Fork[T](input: TypedPipe[T]) extends TypedPipe[T]
    final case class FlatMapValues[K, V, U](input: TypedPipe[(K, V)], fn: V => TraversableOnce[U]) extends TypedPipe[(K, U)]
    final case class FlatMapped[T, U](input: TypedPipe[T], fn: T => TraversableOnce[U]) extends TypedPipe[U]
-   final case class ForceToDisk[T](pipe: TypedPipe[T]) extends TypedPipe[T]
+   final case class ForceToDisk[T](input: TypedPipe[T]) extends TypedPipe[T]
+   final case class Fork[T](input: TypedPipe[T]) extends TypedPipe[T]
+   final case class HashCoGroup[K, V, W, R](left: TypedPipe[(K, V)], right: HashJoinable[K, W], joiner: (K, V, Iterable[W]) => Iterator[R]) extends TypedPipe[(K, R)]
    final case class IterablePipe[T](iterable: Iterable[T]) extends TypedPipe[T]
    final case class MapValues[K, V, U](input: TypedPipe[(K, V)], fn: V => U) extends TypedPipe[(K, U)]
    final case class Mapped[T, U](input: TypedPipe[T], fn: T => U) extends TypedPipe[U]
    final case class MergedTypedPipe[T](left: TypedPipe[T], right: TypedPipe[T]) extends TypedPipe[T]
+   final case class ReduceStepPipe[K, V1, V2](reduce: ReduceStep[K, V1, V2]) extends TypedPipe[(K, V2)]
    final case class SourcePipe[T](source: TypedSource[T]) extends TypedPipe[T]
    final case class SumByLocalKeys[K, V](input: TypedPipe[(K, V)], semigroup: Semigroup[V]) extends TypedPipe[(K, V)]
-   final case class TrappedPipe[T, U >: T](input: TypedPipe[T], sink: Source with TypedSink[T], conv: TupleConverter[U]) extends TypedPipe[U]
+   final case class TrappedPipe[T](input: TypedPipe[T], sink: Source with TypedSink[T], conv: TupleConverter[T]) extends TypedPipe[T]
    final case class WithDescriptionTypedPipe[T](input: TypedPipe[T], description: String, deduplicate: Boolean) extends TypedPipe[T]
    final case class WithOnComplete[T](input: TypedPipe[T], fn: () => Unit) extends TypedPipe[T]
+
    case object EmptyTypedPipe extends TypedPipe[Nothing]
-   final case class HashCoGroup[K, V, W, R](left: TypedPipe[(K, V)],
-    right: HashJoinable[K, W],
-    joiner: (K, V, Iterable[W]) => Iterator[R]) extends TypedPipe[(K, R)]
 
-   final case class CoGroupedPipe[K, V](cogrouped: CoGrouped[K, V]) extends TypedPipe[(K, V)]
-   final case class ReduceStepPipe[K, V1, V2](reduce: ReduceStep[K, V1, V2]) extends TypedPipe[(K, V2)]
+   implicit class InvariantTypedPipe[T](val pipe: TypedPipe[T]) extends AnyVal {
```

Comment:
I'm not clear as to why the implicit class is needed here. Why not leave as a method in the TypedPipe trait (as before, just with the `T` param)?

Reply:
Since `TypedPipe[+T]` is covariant, and so it `TupleConverter[+T]`, you can't accept a TupleConverter as a parameter to TypedPipe. That is why we had `[T, U >: T]` and took `TupleConverter[U]`. However, this really complicates the case class on matches.

Generally, I like this trick: put methods that require such complexities on `implicit class InvariantFoo` so you don't over complicate the code with `[T, U >: T]` type signatures.

Reply:
Ah ok, thanks for clarifying.